### PR TITLE
Add a TensorFlow kernel that wraps WarpCTC

### PR DESCRIPTION
**Pull request by jaredcasper**
Includes a breaking change to the interface.  Instead of ctcComputeInfo there is ctcOptions which allows the caller to specify which label is the reserved blank label.  This was needed because TensorFlow's CTCLoss op has the blank label as the highest value, not zero.

Thus, the current HEAD should be tagged (probably "v1.0"?) and a branch created ("v1"?) from there before this is merged.

### CODE REVIEW

**Comment by jaredcasper** [comment]
@shubho will you take a look?

File: tensorflow_binding/setup.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          file=sys.stderr)
+    sys.exit(1)
+
+warp_ctc_path = "../build"
+if "WARP_CTC_PATH" in os.environ:
+    warp_ctc_path = os.environ["WARP_CTC_PATH"]
+if not os.path.exists(os.path.join(warp_ctc_path, "libwarpctc.so")):
+    print(("Could not find libwarpctc.so in {}.\n"
+           "Build warp-ctc and set WARP_CTC_PATH to the location of"
+           " labwarpctc.so (default is '../build')").format(warp_ctc_path),
```
Comment (bryancatanzaro) [inline tensorflow_binding/setup.py]: should be `libwarpctc.so`

File: tensorflow_binding/setup.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import unittest
+from setuptools.command.build_ext import build_ext as orig_build_ext
+
+# We need to import tensorflow to find where its include directory is.
+try:
+    import tensorflow as tf
+except ImportError:
+    raise RuntimeError("Tensorflow must be installed to build the tensorflow wrapper.")
+
+if "CUDA_HOME" not in os.environ:
```
Comment (bryancatanzaro) [inline tensorflow_binding/setup.py]: Many users might have CPU-only boxes. Might be worth considering whether this should work without a GPU.

File: tests/test_gpu.cu [hunk 3]
```diff
@@ -330,9 +501,10 @@ int main(void) {
     throw_on_error(cudaSetDevice(0), "cudaSetDevice");
 
     bool status = true;
-    status &= small_test();
-    status &= inf_test();
-    status &= run_tests();
+    //status &= small_test();
```
Comment (shubho) [inline tests/test_gpu.cu]: Why are `small_test`, `inf_test` and `run_tests` commented out?

File: tensorflow_binding/warpctc_tensorflow/__init__.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#import warpctc_tensorflow.kernels
```
Comment (shubho) [inline tensorflow_binding/warpctc_tensorflow/__init__.py]: Remove comment above?

**Comment by shubho** [comment]
My comments are mostly nits - except for some commented out tests and it would be nice to be able to install and run on CUDA less machines - not sure how hard that is.

File: tests/test_gpu.cu [hunk 5]
```diff
@@ -46,28 +47,33 @@ bool small_test() {
 
     float score;
 
-    ctcComputeInfo info;
-    info.loc = CTC_GPU;
-    info.stream = stream;
+    float *grads_gpu;
```
Comment (shubho) [inline tests/test_gpu.cu]: Since the test only checks the score why is `grads_gpu` needed?

File: tests/test_gpu.cu [hunk 6]
```diff
@@ -86,6 +102,159 @@ bool small_test() {
...
+                                    options),
+                   "Error: compute_ctc_loss in options_test");
+
+    std::vector<float> grads(alphabet_size * T * minibatch);
+    throw_on_error(cudaMemcpyAsync(grads.data(), grads_gpu,
+                                   grads.size() * sizeof(float),
+                                   cudaMemcpyDeviceToHost, stream),
+                   "cudaMemcpyAsync");
+    throw_on_error(cudaStreamSynchronize(stream), "cudaStreamSynchronize");
+
```
Comment (shubho) [inline tests/test_gpu.cu]: `grads_gpu` is never freed

**Pull request merged**